### PR TITLE
Code generation for GCE compute interface

### PR DESCRIPTION
**Pull request by bowei**
Use code generation to "write" most of the GCE cloud provider library. This enables the following:

- Consistent interfaces, including handling of the different API versions (GA, alpha, beta)
- Efficient implementation of cross cutting features such as metrics, logging, tracing etc. Adding such features has in the past been a tedious and error prone endeavor. 
- High fidelity mocks for all of the compute API. What this means is that most of our controller logic can be tested as unit tests in a consistent way without creating individual mocks by hand.

```release-note
NONE
```

### CODE REVIEW

File: pkg/cloudprovider/providers/gce/cloud/filter/filter_test.go [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+		{f: Regexp("s", "abc").AndEqualBool("b", true), o: &S{S: "abc", B: true}, want: true},
+		{f: Regexp("s", "abc").And(EqualBool("b", true)), o: &S{S: "abc", B: true}, want: true},
+		{f: Regexp("unhandled", "xyz"), o: &S{}},
+		{f: Regexp("nested_field.x", "xyz"), o: &S{}},
+		{f: Regexp("nested_field.x", "xyz"), o: &S{NestedField: &inner{"xyz"}}, want: true},
+		{f: NotRegexp("nested_field.x", "xyz"), o: &S{NestedField: &inner{"xyz"}}},
+		{f: Regexp("nested_field.y", "xyz"), o: &S{NestedField: &inner{"xyz"}}},
+		{f: Regexp("nested_field", "xyz"), o: &S{NestedField: &inner{"xyz"}}},
+	} {
+		got := tc.f.Match(tc.o)
```
Comment (nicksardo) [inline pkg/cloudprovider/providers/gce/cloud/filter/filter_test.go]: Not a big deal, but I prefer that test cases are ran as subtests.

**Comment by bowei** [comment]
Any feedback?

File: pkg/cloudprovider/providers/gce/cloud/gen.go [hunk 2]
```diff
@@ -0,0 +1,10446 @@
...
+	BetaAddresses() BetaAddresses
+	GlobalAddresses() GlobalAddresses
+	BackendServices() BackendServices
+	AlphaBackendServices() AlphaBackendServices
+	AlphaRegionBackendServices() AlphaRegionBackendServices
+	Disks() Disks
+	AlphaDisks() AlphaDisks
+	AlphaRegionDisks() AlphaRegionDisks
+	Firewalls() Firewalls
+	ForwardingRules() ForwardingRules
```
Comment (cheftako) [inline pkg/cloudprovider/providers/gce/cloud/gen.go]: Where do we maintain the mapping/docs from GCE concepts (ForwardingRules) or K8S concepts (LoadBalancer)?

File: pkg/cloudprovider/providers/gce/cloud/gen/main.go [hunk 3]
```diff
@@ -0,0 +1,1142 @@
...
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Generator for GCE compute wrapper code. You must regenerate the code after
+// modifying this file:
+//
+//   $ go run gen/main.go > gen.go
```
Comment (yujuhong) [inline pkg/cloudprovider/providers/gce/cloud/gen/main.go]: Maybe adding a script (`hack/verify-xxx.sh`) to ensure the generated code is in sync with the committed input/generation code?

File: pkg/cloudprovider/providers/gce/cloud/gen.go [hunk 4]
```diff
@@ -0,0 +1,10446 @@
...
+}
+
+// ToAlpha retrieves the given version of the object.
+func (m *MockAddressesObj) ToAlpha() *alpha.Address {
+	if ret, ok := m.Obj.(*alpha.Address); ok {
+		return ret
+	}
+	// Convert the object via JSON copying to the type that was requested.
+	ret := &alpha.Address{}
+	if err := copyViaJSON(ret, m.Obj); err != nil {
```
Comment (cheftako) [inline pkg/cloudprovider/providers/gce/cloud/gen.go]: Would be nice to use Converter.Convert for this. However we may want to do that in a follow up PR.

File: hack/.golint_failures [hunk 5]
```diff
@@ -86,6 +86,7 @@ pkg/cloudprovider
 pkg/cloudprovider/providers/aws
 pkg/cloudprovider/providers/fake
 pkg/cloudprovider/providers/gce
+pkg/cloudprovider/providers/gce/cloud
```
[the remark this answers is not in the document]
Comment (bowei) [reply hack/.golint_failures]: Compute API naming conventions do not match Golang CAPITALIZATIONs.

I fixed all other instances of lint errors.

File: pkg/cloudprovider/providers/gce/cloud/gen.go (hunk 2 above)
Comment (bowei) [reply pkg/cloudprovider/providers/gce/cloud/gen.go]: That is a TODO that needs to be done (but probably can't be addressed here...)

File: pkg/cloudprovider/providers/gce/cloud/gen.go [hunk 6]
```diff
@@ -0,0 +1,10446 @@
...
+	"k8s.io/kubernetes/pkg/cloudprovider/providers/gce/cloud/meta"
+
+	alpha "google.golang.org/api/compute/v0.alpha"
+	beta "google.golang.org/api/compute/v0.beta"
+	ga "google.golang.org/api/compute/v1"
+)
+
+// Cloud is an interface for the GCE compute API.
+type Cloud interface {
+	Addresses() Addresses
```
Comment (yujuhong) [inline pkg/cloudprovider/providers/gce/cloud/gen.go]: Just a thought, maybe it's worth always adding "regional/region" or "zonal/zone" in the function names for consistency (at the cost of slightly longer names). It makes the function names more self-contained by themselves (and not relying on extra knowledge of the GCE resources)

**Review by cheftako — APPROVED** [review]
/lgtm

File: pkg/cloudprovider/providers/gce/cloud/gen.go (hunk 6 above)
Comment (bowei) [reply pkg/cloudprovider/providers/gce/cloud/gen.go]: Not sure about this -- as the compute API already adds "Global", "Region" in inconsistent ways so it'll look even more strange (GlobalAddresses, RegionDisk, ...)

**Review by G-Harmon — COMMENTED** [review]
I looked at the example in https://github.com/bowei/gce-gen/blob/master/cmd/example/main.go and it looks pretty straight-forward.

When you talked about mocks, I was thinking that meant there's a way to verify the behavior of the code that uses the mock. That doesn't seem to be the case. Did I miss that capability?

How does the code-gen know about functions to create other than the basic CRUD operations? For example  (g *GCEBackendServices) GetHealth

**Comment by madhusudancs** [comment]
/approve

(for `hack/`)

**Comment by ixdy** [comment]
Two thoughts:
* We want to move away from checking in generated code (#8830). This one is simple enough that the existing Makefile and Bazel systems should support this easily.
* +1 to @yujuhong's [comment](https://github.com/kubernetes/kubernetes/pull/57879#discussion_r160241921) - if we're going to check in generated code, we need corresponding update/verify scripts to go with it, as otherwise it will inevitably drift out-of-date.

**Comment by bowei** [comment]
@ixdy -- how do we handle no checking in generated code for situations where the generated code is going to be vendored into other projects?

**Comment by ixdy** [comment]
That's still somewhat of an open question. It's one thing that Bazel would do well, assuming both projects are using Bazel. But basically it would be up to the project importing the generated code to make sure the right generation steps were followed when building.

**Comment by ixdy** [comment]
/lgtm cancel

pending hack scripts

**Comment by nicksardo** [comment]
/lgtm
(after offline convo between bowie and ixdy)

**Pull request merged**